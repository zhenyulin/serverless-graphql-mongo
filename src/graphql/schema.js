import schemaComposer from './resolver';

export default schemaComposer.buildSchema();
