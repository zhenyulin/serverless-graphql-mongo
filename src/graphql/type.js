export default `
	type User {
		_id: ID!
		name: String
		email: String!
		followees: [User]
		followers: [User]
		likeItems: [Item]
	}

	input UserInput {
		name: String
		email: String
	}

	type Item {
		_id: ID!
		name: String!
		likedByUsers: [User]
	}

	input ItemInput {
		name: String
	}

	type Query {
		User(_id: ID!): User
		Users: [User]
		Item(_id: ID!): Item
		Items: [Item]
	}

	type Mutation {
		CreateUser(user: UserInput): User
		UpdateUser(_id: ID!, user: UserInput): User
		DeleteUser(_id: ID!): User
		CreateItem(item: ItemInput): Item
		UpdateItem(_id: ID!, item: ItemInput): Item
		DeleteItem(_id: ID!): Item
		UserFollowUser(followerId: ID!, followeeId: ID!): User
		UserUnfollowUser(followerId: ID!, followeeId: ID!): User
		UserLikeItem(userId: ID!, itemId: ID!): User
		UserDisplikeItem(userId: ID!, itemId: ID!): User
	}
`;
