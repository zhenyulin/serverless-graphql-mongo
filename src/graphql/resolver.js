import Mongo from 'models';

export default {
	Query: {
		User: (_, { _id }) => Mongo.User.findById(_id),
		Users: () =>
			Mongo.User.find()
				.setOptions({ sort: { _id: 1 } })
				.lean(),
		Item: (_, { _id }) => Mongo.Item.findById(_id),
		Items: () =>
			Mongo.Item.find()
				.setOptions({ sort: { _id: 1 } })
				.lean(),
	},
	User: {
		followees: async ({ _id }) => {
			const userFollows = await Mongo.UserFollowUser.find({
				followerId: _id,
			});
			const followeeIds = userFollows.map(follow => follow.followeeId);
			const followees = await Mongo.User.find({ _id: { $in: followeeIds } });
			return followees;
		},
		followers: async ({ _id }) => {
			const userFollowed = await Mongo.UserFollowUser.find({
				followeeId: _id,
			});
			const followerIds = userFollowed.map(follow => follow.followerId);
			const followers = await Mongo.User.find({ _id: { $in: followerIds } });
			return followers;
		},
		likeItems: async ({ _id }) => {
			const likes = await Mongo.UserLikeItem.find({
				userId: _id,
			});
			const itemIds = likes.map(like => like.itemId);
			const items = await Mongo.Item.find({ _id: { $in: itemIds } });
			return items;
		},
	},
	Item: {
		likedByUsers: async ({ _id }) => {
			const likes = await Mongo.UserLikeItem.find({
				itemId: _id,
			});
			const userIds = likes.map(like => like.userId);
			const users = await Mongo.User.find({ _id: { $in: userIds } });
			return users;
		},
	},
	Mutation: {
		CreateUser: (_, { user }) => Mongo.User.create(user),
		UpdateUser: (_, { _id, user }) =>
			Mongo.User.findByIdAndUpdate(_id, user, { new: true }),
		DeleteUser: (_, { _id }) => Mongo.User.findOneAndDelete({ _id }),
		CreateItem: (_, { item }) => Mongo.Item.create(item),
		UpdateItem: (_, { _id, item }) =>
			Mongo.Item.findByIdAndUpdate(_id, item, { new: true }),
		DeleteItem: (_, { _id }) => Mongo.Item.findOneAndDelete({ _id }),
		UserFollowUser: async (_, { followerId, followeeId }) => {
			await Mongo.UserFollowUser.findOrCreate({
				followerId,
				followeeId,
			});
			const user = await Mongo.User.findById(followerId);
			return user;
		},
		UserUnfollowUser: async (_, { followerId, followeeId }) => {
			await Mongo.UserFollowUser.findOneAndDelete({ followerId, followeeId });
			const user = await Mongo.User.findById(followerId);
			return user;
		},
		UserLikeItem: async (_, { userId, itemId }) => {
			await Mongo.UserLikeItem.findOrCreate({
				userId,
				itemId,
			});
			const user = await Mongo.User.findById(userId);
			return user;
		},
		UserDislikeItem: async (_, { userId, itemId }) => {
			await Mongo.UserLikeItem.findOneAndDelete({ userId, itemId });
			const user = await Mongo.User.findById(userId);
			return user;
		},
	},
};
