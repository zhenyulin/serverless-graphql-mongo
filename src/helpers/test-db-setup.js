import mongoose from 'mongoose';
import { Mockgoose } from 'mockgoose';

import DB from 'models';

import * as mock from '__fixtures__';

const mockgoose = new Mockgoose(mongoose);

export const connect = async () => {
	await mockgoose.prepareStorage();
	await mongoose.connect(
		process.env.MONGODB_URI,
		{
			useNewUrlParser: true,
		},
	);
};

export const disconnect = () => mongoose.connection.close();

export const loadUsers = () => DB.User.create(mock.USERS);
export const loadItems = () => DB.Item.create(mock.ITEMS);
export const loadUserFollowUsers = () =>
	DB.UserFollowUser.create(mock.USER_FOLLOW_USERS);
export const loadUserLikeItems = () =>
	DB.UserLikeItem.create(mock.USER_LIKE_ITEMS);

export const unloadUsers = () => DB.User.remove();
export const unloadItems = () => DB.Item.remove();
export const unloadUserFollowUsers = () => DB.UserFollowUser.remove();
export const unloadUserLikeItems = () => DB.UserLikeItem.remove();

export const loadAll = () =>
	Promise.all([
		loadUsers(),
		loadItems(),
		loadUserFollowUsers(),
		loadUserLikeItems(),
	]);

export const unloadAll = () =>
	Promise.all([
		unloadUsers(),
		unloadItems(),
		unloadUserFollowUsers(),
		unloadUserLikeItems(),
	]);
