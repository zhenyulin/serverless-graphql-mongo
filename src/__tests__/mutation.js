import { graphql } from 'graphql';

import schema from 'graphql/schema';
import Mongo from 'models';

import * as testDB from './helpers/test-db-setup';

const testQuery = async query => {
	const result = await graphql(schema, query);
	expect(result).toMatchSnapshot();
};

describe('Mutation', () => {
	beforeAll(async () => {
		await testDB.connect();
	});

	afterAll(async () => {
		await testDB.disconnect();
	});

	beforeEach(async () => {
		await testDB.unloadAll();
		await testDB.loadAll();
	});

	describe('CreateUser', () => {
		it('return the created user', async () => {
			const initState = await Mongo.User.find({
				email: 'jack.daniel@test.com',
			});
			expect(initState).toHaveLength(0);
			await testQuery(`
				mutation {
					CreateUser(user: { name: "Jack Daniel", email: "jack.daniel@test.com" }) {
						name
					}
				}
			`);
			const result = await Mongo.User.find({ email: 'jack.daniel@test.com' });
			expect(result).toHaveLength(1);
		});

		it('throws error if email already existed', async () => {
			await testQuery(`
				mutation {
					CreateUser(user: { name: "Jack Daniel", email: "john.cage@test.com" }) {
						name
						email
					}
				}
			`);
		});

		it('throws error if email is missing', async () => {
			await testQuery(`
				mutation {
					CreateUser(user: { name: "Jack Daniel" }) {
						name
						email
					}
				}
			`);
		});
	});

	describe('UpdateUser', () => {
		it('return the updated user', async () => {
			await testQuery(`
				mutation {
					UpdateUser(_id: "4edd40c86762e0fb12000004", user: { name: "John Lemon" }){
						_id
						name
					}
				}
			`);
			const result = await Mongo.User.findById('4edd40c86762e0fb12000004');
			expect(result.name).toBe('John Lemon');
		});
	});

	describe('DeleteUser', () => {
		it('return the deleted user', async () => {
			await testQuery(`
				mutation {
					DeleteUser(_id: "4edd40c86762e0fb12000004"){
						_id
						name
					}
				}
			`);
			const result = await Mongo.User.findById('4edd40c86762e0fb12000004');
			expect(result).toBe(null);
		});
	});

	describe('CreateItem', () => {
		it('return the created item', async () => {
			await testQuery(`
				mutation {
					CreateItem(item: { name: "2046" }){
						name
					}
				}
			`);
			const result = await Mongo.Item.find({ name: '2046' });
			expect(result).toHaveLength(1);
		});

		it('throw error if name is missing', async () => {
			await testQuery(`
				mutation {
					CreateItem(item: {}){
						name
					}
				}
			`);
		});
	});

	describe('UpdateItem', () => {
		it('return the updated item', async () => {
			await testQuery(`
				mutation {
					UpdateItem(_id: "4edd40c86762e0fb12000013", item: { name: "Working Class Hero" }){
						_id
						name
					}
				}
			`);
			const result = await Mongo.Item.findById('4edd40c86762e0fb12000013');
			expect(result.name).toBe('Working Class Hero');
		});
	});

	describe('DeleteItem', () => {
		it('return the deleted item', async () => {
			await testQuery(`
				mutation {
					DeleteItem(_id: "4edd40c86762e0fb12000013"){
						_id
						name
					}
				}
			`);
			const result = await Mongo.Item.findById('4edd40c86762e0fb12000013');
			expect(result).toBe(null);
		});
	});

	describe('UserFollowUser', () => {
		it('return the updated follower', async () => {
			await testQuery(`
				mutation {
					UserFollowUser(followerId: "4edd40c86762e0fb12000005", followeeId: "4edd40c86762e0fb12000004"){
						_id
						name
						followees {
							name
						}
					}
				}
			`);
		});

		it('will not create duplicated follow connections', async () => {
			const query = `
				mutation {
					UserFollowUser(followerId: "4edd40c86762e0fb12000005", followeeId: "4edd40c86762e0fb12000004"){
						_id
						name
						followees {
							name
						}
					}
				}
			`;
			await graphql(schema, query);
			await graphql(schema, query);
			const connections = await Mongo.UserFollowUser.find({
				followerId: '4edd40c86762e0fb12000005',
				followeeId: '4edd40c86762e0fb12000004',
			});
			expect(connections).toHaveLength(1);
		});
	});

	describe('UserUnfollowUser', () => {
		it('return the updated follower', async () => {
			await testQuery(`
				mutation {
					UserUnfollowUser(followerId: "4edd40c86762e0fb12000004", followeeId: "4edd40c86762e0fb12000003"){
						_id
						name
						followees {
							name
						}
					}
				}
			`);
			const connections = await Mongo.UserFollowUser.find({
				followerId: '4edd40c86762e0fb12000004',
				followeeId: '4edd40c86762e0fb12000003',
			});
			expect(connections).toHaveLength(0);
		});
	});

	describe('UserLikeItem', () => {
		it('return the updated user', async () => {
			await testQuery(`
				mutation {
					UserLikeItem(userId: "4edd40c86762e0fb12000005", itemId: "4edd40c86762e0fb12000013"){
						_id
						name
						likeItems {
							name
						}
					}
				}
			`);
		});

		it('will not create duplicated like connections', async () => {
			const query = `
				mutation {
					UserLikeItem(userId: "4edd40c86762e0fb12000005", itemId: "4edd40c86762e0fb12000013"){
						_id
						name
						likeItems {
							name
						}
					}
				}
			`;
			await graphql(schema, query);
			await graphql(schema, query);
			const connections = await Mongo.UserLikeItem.find({
				userId: '4edd40c86762e0fb12000005',
				itemId: '4edd40c86762e0fb12000013',
			});
			expect(connections).toHaveLength(1);
		});
	});

	describe('UserDislikeItem', () => {
		it('return the updated user', async () => {
			const initialState = await Mongo.UserLikeItem.find({
				userId: '4edd40c86762e0fb12000004',
				itemId: '4edd40c86762e0fb12000014',
			});
			expect(initialState).toHaveLength(1);
			await testQuery(`
				mutation {
					UserDislikeItem(userId: "4edd40c86762e0fb12000004", itemId: "4edd40c86762e0fb12000014"){
						_id
						name
						likeItems {
							name
						}
					}
				}
			`);
			const connections = await Mongo.UserLikeItem.find({
				userId: '4edd40c86762e0fb12000004',
				itemId: '4edd40c86762e0fb12000014',
			});
			expect(connections).toHaveLength(0);
		});
	});
});
