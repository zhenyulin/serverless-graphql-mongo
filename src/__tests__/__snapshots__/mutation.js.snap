// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation CreateItem return the created item 1`] = `
Object {
  "data": Object {
    "CreateItem": Object {
      "name": "2046",
    },
  },
}
`;

exports[`Mutation CreateItem throw error if name is missing 1`] = `
Object {
  "data": Object {
    "CreateItem": null,
  },
  "errors": Array [
    [GraphQLError: Item validation failed: name: Path \`name\` is required.],
  ],
}
`;

exports[`Mutation CreateUser return the created user 1`] = `
Object {
  "data": Object {
    "CreateUser": Object {
      "name": "Jack Daniel",
    },
  },
}
`;

exports[`Mutation CreateUser throws error if email already existed 1`] = `
Object {
  "data": Object {
    "CreateUser": null,
  },
  "errors": Array [
    [GraphQLError: E11000 duplicate key error dup key: { : "john.cage@test.com" }],
  ],
}
`;

exports[`Mutation CreateUser throws error if email is missing 1`] = `
Object {
  "data": Object {
    "CreateUser": null,
  },
  "errors": Array [
    [GraphQLError: User validation failed: email: Path \`email\` is required.],
  ],
}
`;

exports[`Mutation DeleteItem return the deleted item 1`] = `
Object {
  "data": Object {
    "DeleteItem": Object {
      "_id": "4edd40c86762e0fb12000013",
      "name": "Imagine",
    },
  },
}
`;

exports[`Mutation DeleteUser return the deleted user 1`] = `
Object {
  "data": Object {
    "DeleteUser": Object {
      "_id": "4edd40c86762e0fb12000004",
      "name": "John Lennon",
    },
  },
}
`;

exports[`Mutation UpdateItem return the updated item 1`] = `
Object {
  "data": Object {
    "UpdateItem": Object {
      "_id": "4edd40c86762e0fb12000013",
      "name": "Working Class Hero",
    },
  },
}
`;

exports[`Mutation UpdateUser return the updated user 1`] = `
Object {
  "data": Object {
    "UpdateUser": Object {
      "_id": "4edd40c86762e0fb12000004",
      "name": "John Lemon",
    },
  },
}
`;

exports[`Mutation UserDislikeItem return the updated user 1`] = `
Object {
  "data": Object {
    "UserDislikeItem": Object {
      "_id": "4edd40c86762e0fb12000004",
      "likeItems": Array [],
      "name": "John Lennon",
    },
  },
}
`;

exports[`Mutation UserFollowUser return the updated follower 1`] = `
Object {
  "data": Object {
    "UserFollowUser": Object {
      "_id": "4edd40c86762e0fb12000005",
      "followees": Array [
        Object {
          "name": "John Lennon",
        },
      ],
      "name": "Miles Davis",
    },
  },
}
`;

exports[`Mutation UserLikeItem return the updated user 1`] = `
Object {
  "data": Object {
    "UserLikeItem": Object {
      "_id": "4edd40c86762e0fb12000005",
      "likeItems": Array [
        Object {
          "name": "Imagine",
        },
      ],
      "name": "Miles Davis",
    },
  },
}
`;

exports[`Mutation UserUnfollowUser return the updated follower 1`] = `
Object {
  "data": Object {
    "UserUnfollowUser": Object {
      "_id": "4edd40c86762e0fb12000004",
      "followees": Array [],
      "name": "John Lennon",
    },
  },
}
`;
